doc/internal/man7/EVP_PKEY.pod

```diff
@@ -8,37 +8,198 @@ EVP_PKEY - an internal description
-form, this key is commonly referred to as the "origin".
-
-An B<EVP_PKEY> also contains a cache of provider side copies of the
-key, each adapted for the provider that is going to use that copy to
-perform some operation.
-For a legacy "origin", the B<EVP_PKEY_ASN1_METHOD>'s functions
-export_to() and dirty_cnt() must be implemented for such caching to be
-possible.  For a provider side "origin", the B<EVP_KEYMGMT>'s function
-OP_keymgmt_export() must be implemented.  In all cases, the receiving
-B<EVP_KEYMGMT> must have an implemented OP_keygmt_import().
+The private/public key pair that an B<EVP_PKEY> contains is refered to
+as its "internal key" or "origin" (the reason for "origin" is
+explained further down, in L</Export cache for provider operations>),
+and it can take one of the following forms:
+
+=over 4
+
+=item legacy origin
+
+This is the form that an B<EVP_PKEY> in OpenSSL prior to 3.0 had.  The
+internal key in the B<EVP_PKEY> is a pointer to to the low-level key
+types, such as B<RSA>, B<DSA> and B<EC>, or an engine driven
+structure, and is governed by an associated L<EVP_PKEY_METHOD(3)> and
+an L<EVP_PKEY_ASN1_METHOD(3)>.
+
+The functions available through those two method structures get full
+access to the B<EVP_PKEY> and therefore have a lot of freedom to
+modify whatever they want.  This also means that an B<EVP_PKEY> is a
+shared structure between libcrypto and any ENGINE that serves such
+methods.
+
+=item provider-native origin
+
+This is a new form in OpenSSL 3.0, which permits providers to hold the
+key data (see L<provider-keymgmt(7)>).  The internal key in the
+B<EVP_PKEY> is a pointer to that key data held by the provider, and
+is governed by an associated L<EVP_KEYMGMT(3)> method structure.
+
+The functions available through The L<EVP_KEYMGMT(3)> have no access
+to the B<EVP_PKEY>, and can therefore not make any direct changes.
+Similarly, the key data that the B<EVP_PKEY> points at is only known
+to the functions pointed at in the L<EVP_KEYMGMT(3)>.
+
+=back
+
+These two forms can never co-exist in the same B<EVP_PKEY>, the main
+reason being that having both at the same time will create problems
+with synchronising between the two forms, and potentially make it
+confusing which one of the two is the origin.
+
+=head2 Key mutability
+
+The B<EVP_PKEY> internal keys are mutable.
+
+This is especially visible with internal legacy keys, since they can
+be extracted with functions like L<EVP_PKEY_get0_RSA(3)> and then
+modified at will with functions like L<RSA_set0_key(3)>.
+
+Internal provider native keys are also possible to be modified, if the
+associated L<EVP_KEYMGMT(3)> implementation allows it.  This is done
+with L<EVP_PKEY_set_params(3)> and its specialised derivatives.  The
+OpenSSL providers allow it for the following:
+
+=over 4
+
+=item DH, EC, ED25519, ED448, X25519, X448:
+
+It's possible to set the encoded public key.  This is supported in
+particular through L<EVP_PKEY_set1_encoded_public_key(3)>.
```

Comment:
That manpage says it's only for key exchanges, so I guess ED* should be removed.

Does this mean you need to create a new EVP_PKEY for each key exchange? It doesn't seem thread safe.

Reply:
Thanks, I've removed "ED25519, ED448" after verifying that the code indeed limits what key types are affected.

> Does this mean you need to create a new EVP_PKEY for each key exchange? It doesn't seem thread safe.

I'm not sure I understand the question...

Reply:
I'm looking at the documentation, and I can't actually figure out how it works. I want to do a DH key exchange, and there is a `p` and `g`. Does that get set in pkey in the example in EVP_PKEY_derive.pod? The peerkey, which you can set using EVP_PKEY_set1_encoded_public_key() ends up in a different EVP_PKEY? Does that also have `p` and `g` set in it?

Reply:
Er...  if you do `EVP_PKEY_set1_encoded_public_key(pk, pub, publen)`, the public key given by `pub` will end up in the internal origin key of `pk`.  This will also trip the dirty count, which will result in an invalidated cache later on.

Reply:
I think I understand the underlying request, to write a little more about what happens when a key is modified.

Reply:
Now that I'm reading what I've written again, I'm not sure what to add to clear things up.  I was hoping that the "Key mutability" section (which is what we're commenting here) answered most question on what happens when you make changes to the internal key.

The story about a peer key related to `EVP_PKEY_set1_encoded_public_key()`, I'm not sure I understand.  The description for this function starts with this:

> **EVP_PKEY_set1_encoded_public_key()** can be used to set the public key value within an existing EVP_PKEY object.

This is to be understood that it's the internal origin key that is changed, and "Key mutability" tells us that this trips the dirty count.

If we can clear that in some manner, that would be nice, but in the mean time, I'm not sure if this is a blocker for this PR.

Reply:
You actually should create a new EVP_PKEY for the peer key for each key exchange. I do not see that as a limitation as you usually do not do multiple key exchanges with the same peer at once. We certainly should not try to make simultaneous update access of an EVP_PKEY possible as it was not possible in 1.1.1 either.

Reply:
Since the answer is that you always need to create a new EVP_PKEY, why first create an incomplete key and then modify it? So I'm really asking if that is the API we want, and that the new EVP_PKEY_set1_encoded_public_key() should exist or not. This was functionality that existed in 1.1.1, so we must support it, but maybe only with the deprecated APIs.

The documentation seems correct for what is currently implemented.

Reply:
The problem is that the parameters and the encoded public key usually come from slightly different sources so the way it is designed now is really convenient for the applications. Where if you wanted to do it all via single fromdata() you'd have to first export the parameters to OSSL_PARAMS and then add the key. Also the key is in some encoded form and not a raw bignum so this also makes it slightly conceptually awkward to combine it with the parameters in a single fromdata().

Reply:
I think that the answer is here:

``` console
$ git grep set1_encoded
...
crypto/cms/cms_dh.c:     * EVP_PKEY_set1_encoded_public_key()
crypto/cms/cms_dh.c:            || !EVP_PKEY_set1_encoded_public_key(pkpeer, buf, plen))
crypto/cms/cms_ec.c:    if (!EVP_PKEY_set1_encoded_public_key(pkpeer, p, plen))
...
ssl/statem/extensions_clnt.c:        if (EVP_PKEY_set1_encoded_public_key(skey, PACKET_data(&encoded_pt),
ssl/statem/extensions_srvr.c:        if (EVP_PKEY_set1_encoded_public_key(s->s3.peer_tmp,
ssl/statem/statem_clnt.c:    if (EVP_PKEY_set1_encoded_public_key(s->s3.peer_tmp,
ssl/statem/statem_srvr.c:    if (!EVP_PKEY_set1_encoded_public_key(ckey, data, i)) {
ssl/statem/statem_srvr.c:        if (EVP_PKEY_set1_encoded_public_key(ckey, data, i) <= 0) {
```

I don't have the expertise to say if this is a legitimate thing to do or not, but the alternative would be to create a new copy of the key using an OSSL_PARAM with the encoded public key added.  That seems like a lot of churn.

As for the key being in encoded form, I'm not sure I see the problem...  it's actually not encoded in any special way, it's simply an octet string, and could be transferred using the data type OSSL_PARAM_OCTET_STRING...  which, incidently, is exactly what `EVP_PKEY_set1_encoded_public_key()` does.

Reply:
That being said, this discussion is off topic for this PR.  Maybe it's time to raise an issue on the subject?